### PR TITLE
[datasets-overhaul #2] Add a new Dataset class.

### PR DESCRIPTION
**Pull request by ChrisCummins**
A dataset is a collection of benchmarks.

Datasets provide a convenience abstraction for installing and managing groups of benchmarks. Datasets provide the API for instantiating benchmarks, either by randomly selecting from the available benchmarks or by selecting by URI.

The Dataset class is an abstract base for implementing datasets. At a minimum, subclasses must implement the `benchmark()` and `benchmark_uris()` methods. Other methods such as `install()` may be used where helpful.

Issue #45.

### CODE REVIEW

File: tests/datasets/dataset_test.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+from compiler_gym.datasets.dataset import Dataset
+from tests.test_main import main
+
+pytest_plugins = ["tests.pytest_plugins.common"]
+
+# pylint: disable=abstract-method
+
+
+def test_dataset_properties(tmpwd: Path):
+    del tmpwd  # Unused
```
Comment (JD-ETH) [inline tests/datasets/dataset_test.py]: why is the argument required for the test function?

**Review by JD-ETH — APPROVED** [review]
I think the structure is very clear now this way, don't see issues here.

File: tests/datasets/dataset_test.py (hunk 1 above)
Comment (ChrisCummins) [reply tests/datasets/dataset_test.py]: Good catch! In an early revision, `Dataset` would create a directory on init, so the test needed a scrap working directory to write to. This is no longer required, I'll remove it.

**Comment by ChrisCummins** [comment]
Thanks for the review @JD-ETH!

**Pull request merged**